### PR TITLE
Implement better handling of divergence

### PR DESCRIPTION
**Pull request by flodiebold**
Divergence here means that for some reason, the end of a block will not be reached. We tried to model this just using the never type, but that doesn't work fully (e.g. in `let x = { loop {}; "foo" };` x should still have type `&str`); so this introduces a `diverges` flag that the type checker keeps track of, like rustc does. We also add some checking for `break`, but no support for break-with-value or labeled breaks yet.

### CODE REVIEW

File: crates/ra_hir_ty/src/lib.rs [hunk 1]
```diff
@@ -730,6 +730,13 @@ impl Ty {
         }
     }
 
+    pub fn is_never(&self) -> bool {
+        match self {
```
Comment (matklad) [inline crates/ra_hir_ty/src/lib.rs:734]: `matches!(self, Ty::Apply(ApplicationTy { ctor: TypeCtor::Never, .. }))`

File: crates/ra_hir_ty/src/infer/expr.rs [hunk 2]
```diff
@@ -191,6 +233,9 @@ impl<'a> InferenceContext<'a> {
                     // FIXME handle break with value
                     self.infer_expr(*expr, &Expectation::none());
                 }
+                if let Some(ctxt) = self.breakables.last_mut() {
```
Comment (matklad) [inline crates/ra_hir_ty/src/infer/expr.rs:236]: I guess we might emit a diagnostic here? Seems like we can't do this syntactically, as it needs to happen post expansion

Comment (flodiebold) [reply crates/ra_hir_ty/src/infer/expr.rs:236]: I added the diagnostic. It's quite a bit of boilerplate :sweat:

Comment (matklad) [reply crates/ra_hir_ty/src/infer/expr.rs:236]: Yeah... 

Designing diagnostics infra seems like a task we can tackle now, as we actually have quite a few things we can report w/o false positives now...

Though, I must say I like how concise the emitting part looks

**Pull request merged**